### PR TITLE
Options liberty added

### PR DESCRIPTION
**Pull request by WhiteWLf-dev**

### CODE REVIEW

File: src/gate/optimizer/tech_map/strategy/replacement_cut.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  void ReplacementVisitor::set(CutStorage *cutStorage,
+      GNet *net, 
+      std::unordered_map<GateID, Replacement> *bestReplacement,
+      int cutSize) {
+    this->cutStorage = cutStorage;
+    this->net = net;
+    this->cutSize = cutSize;
+    this->bestReplacement = bestReplacement;
+                            }
```
Comment (mchupilko) [inline src/gate/optimizer/tech_map/strategy/replacement_cut.cpp]: two spaces

File: src/gate/optimizer/tech_map/strategy/replacement_cut.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    lastCuts = &(cutStorage->cuts[node]);
+    return SUCCESS;
+  }
+
+  VisitorFlags ReplacementVisitor::onCut(const Visitor::Cut &cut) {
+    return SUCCESS;
+  }
+
+  VisitorFlags ReplacementVisitor::onNodeEnd(const GateID &) {
+
```
Comment (mchupilko) [inline src/gate/optimizer/tech_map/strategy/replacement_cut.cpp]: remove new line

File: src/gate/optimizer/tech_map/strategy/replacement_cut.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        // Discard trivial cuts.
+      } else if (node == lastNode) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  void ReplacementVisitor::finishTechMap() {
+    std::cout << "Do replace\n";
```
Comment (mchupilko) [inline src/gate/optimizer/tech_map/strategy/replacement_cut.cpp]: Use standard logging or comment it away

File: src/gate/optimizer/tech_map/strategy/replacement_cut.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#include "gate/optimizer/rwdatabase.h"
+#include "gate/optimizer/tech_map/replacement_struct.h"
+#include "gate/optimizer/util.h"
+#include "gate/optimizer/visitor.h"
+
+#include <queue>
+
+namespace eda::gate::optimizer {
+/**
+ * \brief Realization of interface Visitor. Replacing with better cuts.
```
Comment (mchupilko) [inline src/gate/optimizer/tech_map/strategy/replacement_cut.h]: Implementation of...

File: src/gate/optimizer/tech_map/tech_map_visitor.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  void TechMapVisitor::set(CutStorage *cutStorage,
+      GNet *net, 
+      std::unordered_map<GateID, Replacement> *bestReplacement,
+      int cutSize) {
+    this->cutStorage = cutStorage;
+    this->net = net;
+    this->cutSize = cutSize;
+    this->bestReplacement = bestReplacement;}
+  
+
```
Comment (mchupilko) [inline src/gate/optimizer/tech_map/tech_map_visitor.cpp:27]: remove new line

File: src/gate/optimizer/tech_map/tech_map_visitor.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace eda::gate::optimizer {
+/**
+ * \brief Realization of interface Visitor.
+ * \author <a href="mailto:dgaryaev@ispras.ru"></a>
+ */
+
+  class TechMapVisitor : public Visitor {
+  public:
+
```
Comment (mchupilko) [inline src/gate/optimizer/tech_map/tech_map_visitor.h]: remove new line

File: src/gate/optimizer/tech_map/tech_map_visitor.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+#include "gate/optimizer/rwdatabase.h"
+#include "gate/optimizer/tech_map/replacement_struct.h"
+#include "gate/optimizer/util.h"
+#include "gate/optimizer/visitor.h"
+
+#include <queue>
+
+namespace eda::gate::optimizer {
+/**
+ * \brief Realization of interface Visitor.
```
Comment (mchupilko) [inline src/gate/optimizer/tech_map/tech_map_visitor.h]: Implementation of...

File: src/gate/optimizer/tech_map/tech_map_visitor.h [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    CutStorage::Cuts *lastCuts;
+    std::vector<const CutStorage::Cut*> toRemove;
+
+    bool checkValidCut(const Cut &cut);
+
+  protected:
+    GNet *net;
+    GateID lastNode;
+    int cutSize;
+
```
Comment (mchupilko) [inline src/gate/optimizer/tech_map/tech_map_visitor.h:58]: remove new line..

**Review by mchupilko — COMMENTED** [review]
Please make minor fixes. Also, conflicts with main should be resolved.

File: src/gate/library/liberty/translate.cpp [hunk 9]
```diff
@@ -33,7 +33,7 @@ static void fillInputs(
     GModel::GNet &net,
     std::map<size_t, GateId> &inputs) {
   std::vector<size_t> tmp;
-  for (auto [name, wire]: wires) {
+  for (auto &[name, wire]: wires) {
```
Comment (mchupilko) [inline src/gate/library/liberty/translate.cpp]: utopia-eda/src/gate/library/liberty/translate.cpp:36:8: warning: 'auto &' can be declared as 'const auto &' [llvm-qualified-auto]

File: src/gate/library/liberty/translate.cpp [hunk 10]
```diff
@@ -388,7 +388,7 @@ bool translateModuleToGNet(
 void translateDesignToGNet(
     const RTlil::Design *des,
     NetData &vec) {
-  for (auto [IdString, Module]: des->modules_) {
+  for (auto &[IdString, Module]: des->modules_) {
```
Comment (mchupilko) [inline src/gate/library/liberty/translate.cpp]: utopia-eda/src/gate/library/liberty/translate.cpp:391:8: warning: 'auto &' can be declared as 'const auto &' [llvm-qualified-auto]

File: src/gate/model/gnet.cpp [hunk 11]
```diff
@@ -167,7 +167,7 @@ void GNet::setGate(GateId gid, GateSymbol func, const SignalList &inputs) {
 
   gate->setFunc(func);
   gate->setInputs(inputs);
-  assert(gate->invariant());
+  //assert(gate->invariant());
```
Comment (greye) [inline src/gate/model/gnet.cpp:170]: After the recent changes the assert (probably) can be uncommented, needs testing.

File: src/gate/optimizer/tech_map/strategy/replacement_cut.cpp [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    this->cutSize = cutSize;
+    this->bestReplacement = bestReplacement;
+  }
+
+  VisitorFlags ReplacementVisitor::onNodeBegin(const GateID &node) {
+    if (cutStorage->cuts.find(node) == cutStorage->cuts.end()) {
+      // If node is not in cutStorage - means, that it is a new node.
+      // So we recount cuts for that node.
+     CutsFindVisitor finder(cutSize, cutStorage);
+     finder.onNodeBegin(node);
```
Comment (greye) [inline src/gate/optimizer/tech_map/strategy/replacement_cut.cpp:34]: Broken indentation

File: src/gate/optimizer/tech_map/strategy/replacement_cut.cpp [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    lastNode = node;
+    lastCuts = &(cutStorage->cuts[node]);
+    return SUCCESS;
+  }
+
+  VisitorFlags ReplacementVisitor::onCut(const Visitor::Cut &cut) {
+   return SUCCESS;
```
Comment (greye) [inline src/gate/optimizer/tech_map/strategy/replacement_cut.cpp:42]: Broken indentation

File: src/gate/optimizer/tech_map/strategy/replacement_cut.h [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace eda::gate::optimizer {
+/**
+ * \brief Implementation of interface Visitor. Replacing with better cuts.
+ * \author <a href="mailto:dgaryaev@ispras.ru"></a>
+ */
+
+  class ReplacementVisitor : public Visitor {
+  public:
+
```
Comment (greye) [inline src/gate/optimizer/tech_map/strategy/replacement_cut.h:28]: Redundant empty line

File: src/gate/optimizer/tech_map/strategy/replacement_cut.h [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    bool checkValidCut(const Cut &cut);
+    void finishTechMap();
+
+  protected:
+    GNet *net;
+    GateID lastNode;
+    int cutSize;
+    RWDatabase rwdb;
+
```
Comment (greye) [inline src/gate/optimizer/tech_map/strategy/replacement_cut.h:61]: Redundant empty line

File: src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+  BoundGNetList
+  SimpleTechMapper::getSubnets(uint64_t func) {
+    return rwdb.get(func);
+  }
+
+  void SimpleTechMapper::finishTechMap() {
+    if (bestOptionMap.size() != 0) {
+      Replacement bestReplacment{lastNode, bestOptionMap, bestOption.net.get(), 
+        net, minNodeArrivalTime};
```
Comment (greye) [inline src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp:40]: Broken formatting

File: src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+      bestReplacement->insert(std::pair<GateID, Replacement>
+                              (lastNode, bestReplacment));
```
Comment (greye) [inline src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp:42]: Broken formatting

File: src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    }
+  }
+
+  double SimpleTechMapper::maxArrivalTime(const BoundGNet &superGate,
+      const std::unordered_map<GateID, GateID> &map) {
+
+    double maxDelay = 0;
+
```
Comment (greye) [inline src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp:50]: Redundant empty lines

File: src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+    std::unordered_map<Gate::Id, uint32_t> revGareBindings;
+
+    for (const auto &[superInputId, superGateId] : superGate.bindings) {
+        revGareBindings[superGateId] = superInputId;
+      }
```
Comment (greye) [inline src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp:55]: Broken indentation

File: src/gate/optimizer/tech_map/tech_map_visitor.cpp [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+      finder.onNodeBegin(node);
+    }
+    lastNode = node;
+    lastCuts = &(cutStorage->cuts[node]);
+    return SUCCESS;
+  }
+
+  VisitorFlags TechMapVisitor::onCut(const Visitor::Cut &cut) {
+    if (checkValidCut(cut)) {
+
```
Comment (greye) [inline src/gate/optimizer/tech_map/tech_map_visitor.cpp:42]: Redundant empty line

File: src/gate/optimizer/tech_map/tech_map_visitor.cpp [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+      const auto & cutConeMap = coneVisitor.getResultCut();
+      for(const auto &[gateSource, gateCone] : cutConeMap) {
+        boundGNet.bindings[boundGNet.bindings.size()] = gateCone;
+      }
+
+      auto func = TTBuilder::build(boundGNet);
+
+      auto list = getSubnets(func);
+
```
Comment (greye) [inline src/gate/optimizer/tech_map/tech_map_visitor.cpp:60]: Redundant empty lines

File: src/gate/optimizer/tech_map/tech_map_visitor.h [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    virtual bool checkOptimize(const BoundGNet &superGate,
+                               const std::unordered_map<GateID, GateID> &map) = 0;
+
+    virtual VisitorFlags
+    considerTechMap(BoundGNet &superGate,
+                         std::unordered_map<GateID, GateID> &map) = 0;
```
Comment (greye) [inline src/gate/optimizer/tech_map/tech_map_visitor.h:64]: Broken indentation

File: src/gate/optimizer/tech_map/tech_mapper.cpp [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+    CutStorage cutStorage = findCuts(cutSize, net);
+
+    techMapper.set(&cutStorage, net, &bestReplacement, cutSize);
+
+    Walker walker(net, &techMapper, &cutStorage);
+
+    while(true) {
+      walker.walk(true);
+      break;
+    }
```
Comment (greye) [inline src/gate/optimizer/tech_map/tech_mapper.cpp:29]:
```suggestion
    walker.walk(true);
```

File: src/gate/optimizer/tech_map/tech_mapper.cpp [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+
+    if (bestReplacement.size() != 0) {
+      replacer.set(&cutStorage, net, &bestReplacement, cutSize);
+
+      Walker repl(net, &replacer, &cutStorage);
+
+      while(true) {
+        repl.walk(false);
+        break;
+      }
```
Comment (greye) [inline src/gate/optimizer/tech_map/tech_mapper.cpp:39]:
```suggestion
      repl.walk(false);
```

Comment (Scytti) [reply src/gate/optimizer/tech_map/tech_mapper.cpp:39]: Fixed

File: src/gate/optimizer/tech_map/tech_mapper.cpp (hunk 23 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/tech_mapper.cpp:29]: Fixed

File: src/gate/optimizer/tech_map/tech_map_visitor.h (hunk 22 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/tech_map_visitor.h:64]: Fixed

File: src/gate/optimizer/tech_map/tech_map_visitor.cpp (hunk 21 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/tech_map_visitor.cpp:60]: Removed lines

File: src/gate/optimizer/tech_map/tech_map_visitor.cpp (hunk 20 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/tech_map_visitor.cpp:42]: Removed the line

File: src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp (hunk 19 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp:55]: Fixed indentation

File: src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp (hunk 18 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp:50]: Removed lines

File: src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp (hunk 17 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp:42]: Fixed

File: src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp (hunk 16 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/strategy/simple_techmapper.cpp:40]: Fixed

File: src/gate/optimizer/tech_map/strategy/replacement_cut.h (hunk 15 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/strategy/replacement_cut.h:61]: Removed the line

File: src/gate/optimizer/tech_map/strategy/replacement_cut.h (hunk 14 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/strategy/replacement_cut.h:28]: Removed the line

File: src/gate/optimizer/tech_map/strategy/replacement_cut.cpp (hunk 13 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/strategy/replacement_cut.cpp:42]: Fixed indentation

File: src/gate/optimizer/tech_map/strategy/replacement_cut.cpp (hunk 12 above)
Comment (Scytti) [reply src/gate/optimizer/tech_map/strategy/replacement_cut.cpp:34]: Fixed indentation

**Pull request closed without merging**